src/asmcnc/job/job_data.py

```diff
@@ -389,3 +400,209 @@ def post_job_data_update_post_send(self):
+                job_recovery_info_file.write(self.filename + "\n" + str(cancel_line))
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = self.filename
+            self.job_recovery_cancel_line = cancel_line
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not write recovery info")
+            print(str(traceback.format_exc()))
+
+    def clear_recovery_file(self):
+        try:
+            open(self.job_recovery_info_filepath, 'w').close()
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = ''
+            self.job_recovery_cancel_line = 0
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not clear recovery info")
+            print(str(traceback.format_exc()))
+
+
+    def reset_recovery(self):
+        self.job_recovery_selected_line = -1
+        self.job_recovery_gcode = []
+        self.job_recovery_offset = 0
+
+    def generate_recovery_gcode(self):
+        try:
+            recovery_gcode = []
+
+            # Recover modal gcodes
+
+            # Coordinate System Select
+            coord_system_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G5[4-9]", s)), None)
+            if coord_system_line:
+                recovery_gcode.append(re.search("G5[4-9](\.[1-3])?", coord_system_line).group(0))
+
+            # Plane selection
+            plane_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G1[7-9]", s)), None)
+            if plane_line:
+                recovery_gcode.append(re.search("G1[7-9]", plane_line).group(0))
+
+            # Distance mode
+            distance_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]([A-Z]|\s|$)", s)), None)
+            if distance_line:
+                # Recovery not allowed if G91 is used
+                if re.search("G91([A-Z]|\s|$)", distance_line):
+                    return False, 'This job contains the G91 g-code, so it cannot be recovered.'
+                recovery_gcode.append("G90")
+
+            # Arc IJK distance mode
+            arc_mode_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[0,1]\.1", s)), None)
+            if arc_mode_line:
+                recovery_gcode.append(re.search("G9[0,1]\.1", arc_mode_line).group(0))
+
+            # Feed rate mode
+            feedrate_mode_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G9[3-5]", s)), None)
+            if feedrate_mode_line:
+                recovery_gcode.append(re.search("G9[3-5]", feedrate_mode_line).group(0))
```

Comment:
We're not sure how to handle G93 or G95, so return False for these (and error message... @Neex101 any preferences on what that should be?). G94 only for time being :)

Reply:
Error message along lines of: 

"Job recovery only supports feed rate mode G94. This job contains [G93/G95], and therefore cannot be recovered."